Sources/ApodiniHTTP/InterfaceExporter.swift

```diff
@@ -0,0 +1,408 @@
+                    
+                    return Vapor.Response(status: HTTPStatus(status ?? .ok),
+                                          headers: HTTPHeaders(information),
+                                          body: Vapor.Response.Body(data: body))
+                }
+                .firstFuture(on: request.eventLoop)
+                .map { optionalResponse in
+                    optionalResponse ?? Vapor.Response()
+                }
+        }
+    }
+    
+    
+    // MARK: Client Streaming Closure
+    
+    private func buildClientSideStreamingClosure<H: Handler>(
+        for endpoint: Endpoint<H>,
+        using defaultValues: DefaultValueStore) -> (Vapor.Request) throws -> EventLoopFuture<Vapor.Response> {
+        let strategy = multiInputDecodingStrategy(for: endpoint)
+        
+        let abortAnyError = AbortTransformer<H>()
+        
+        let transformer = VaporResponseTransformer<H>(configuration.encoder)
+        
+        return { (request: Vapor.Request) in
+            guard let requestCount = try configuration.decoder.decode(ArrayCount.self, from: request.bodyData).count else {
+                throw ApodiniError(
+                    type: .badInput,
+                    reason: "Expected array at top level of body.",
+                    description: "Input for client side steaming endpoints must be an array at top level.")
+            }
+            
+            var delegate = Delegate(endpoint.handler, .required)
+            
+            return Array(0..<requestCount)
+                .publisher
+                .map { index in
+                    (request, (request, index))
+                }
+                .decode(using: strategy, with: request.eventLoop)
+                .insertDefaults(with: defaultValues)
+                .validateParameterMutability()
+                .cache()
+                .subscribe(to: &delegate)
+                .evaluate(on: &delegate)
+                .transform(using: abortAnyError)
+                .cancel(if: { response in
+                    response.connectionEffect == .close
+                })
+                .compactMap { (response: Apodini.Response<H.Response.Content>) in
+                    if response.connectionEffect == .open && response.content == nil {
+                        return nil
+                    } else {
+                        return response
+                    }
+                }
+                .tryMap { (response: Apodini.Response<H.Response.Content>) -> Vapor.Response in
+                    return try transformer.transform(input: response)
+                }
+                .firstFuture(on: request.eventLoop)
+                .map { optionalResponse in
+                    optionalResponse ?? Vapor.Response()
+                }
+        }
+    }
+    
+    // MARK: Bidirectional Streaming Closure
+    
+    private func buildBidirectionalStreamingClosure<H: Handler>(
```

Comment:
So to my understanding this isn't a real bidirectional streaming that e.g. builds on HTTP/2 but really only multiplexes multiple requests and response into one HTTP/1.1 request response right?

Would using HTTP/2 for that work though? Maybe a great way to showcase that with a example implementation 🤟🏼

Reply:
Yes, correct. Of course, it would make absolute sense to implement versions that make use of HTTP/2 features, but my intention here was more to showcase how the Publisher API can be used to nicely provide a "downgraded" endpoint, where the protocol just doesn't allow for a direct representation. Even if we were to add HTTP/2 implementations for the patterns that can be implemented properly using HTTP/2, we would still require the current implementations for clients that only support HTTP/1.1.

Reply:
Understood.